File: 004/eu004.go
// project euler (projecteuler.net) problem 4
// solution by Kevin Retzke (retzkek@gmail.com) April 2012
package main

import (
	"fmt"
	"math"
	"strconv"
)

// IsPalindrome returns true if num is palindromic.
func isPalindrome(num int) bool {
	str := strconv.Itoa(num)
	for i := 0; i < len(str)/2; i++ {
		if str[i] != str[len(str)-i-1] {
			return false
		}
	}
	return true
}

// Palindrome calculates the largest palindrom made from the product of two
// digits-digit numbers.  The return values are the two numbers and the
// product.
func palindrome(digits int) (int, int, int) {
	var start, stop, num0, num1, largest int
	start = int(math.Pow10(digits)) - 1
	stop = int(math.Pow10(digits - 1))
	for i := start; i >= stop; i-- {
		for j := i; j >= stop; j-- {
			if n := i * j; isPalindrome(n) && n > largest {
				largest = n
				num0 = i
				num1 = j
				fmt.Println(i, j, n)
			}
		}
	}
	return num0, num1, largest
}

func main() {
	var a, b, r int
	// test case
	if a, b, r = palindrome(2); r == 9009 {
		fmt.Println("Test: pass")
	} else {
		fmt.Printf("Test: fail (r=%v (%v * %v))\n", r, a, b)
	}
	// challenge
	a, b, r = palindrome(3)
	fmt.Printf("Result: %v (%v * %v)\n", r, a, b)
}

File: 004/eu004.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problem 4:
solution by Kevin Retzke (retzkek@gmail.com), April 2009

A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 * 99.
Find the largest palindrome made from the product of two 3-digit numbers.
"""

def isPalindrome(num):
    """
    tests if number is palindromic
    """
    str = "%i" % num     # convert number to string
    rev = int(str[::-1]) # reverse and convert back to integer
    return num == rev

def palindrome(digits):
    """
    calculates the largest palindrom made from the product of two digits-digit numbers
    returns the numbers (multiply them to get the palindromic number)
    """
    start = pow(10,digits)-1
    stop = pow(10,digits-1)
    nums=(0,0)
    largest=0

    # naive search multiplying all n-digit numbers and checking if the result is a palindrome
    # note that we need to keep searching even after finding one, since we want the largest,
    # which may not be the first we find (e.g. 999*100 < 998*500)
    for i in range(start,stop-1,-1):
        for j in range(i, stop-1, -1):
            if isPalindrome(i*j) and i*j > largest:
                print "%i (%i * %i)" % (i*j, i, j)
                largest = i*j
                nums = (i,j)
    return nums


if __name__ == '__main__':
    # test case
    (a,b) = palindrome(2)
    if a*b == 9009:
        print "Test: pass"
    else:
        print "Test: fail (numbers = %i, %i; result = %i)" % (a, b, a*b)
    # challenge
    (a,b) = palindrome(3)
    print "Result: %i (%i * %i)" % (a*b, a, b)
